### PR TITLE
Implement some support for JAX-RS 2.1 additions

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
* HTTP method `PATCH`
* Async resource methods returning a `CompletionStage`

### CODE REVIEW

File: instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-common/src/main/java/io/opentelemetry/instrumentation/auto/jaxrs/v2_0/JaxRsAnnotationsInstrumentation.java [hunk 1]
```diff
@@ -155,15 +157,23 @@ public static void stopSpan(
...
+          returnValue instanceof CompletionStage ? (CompletionStage<?>) returnValue : null;
+
       if (asyncResponse != null && !asyncResponse.isSuspended()) {
         // Clear span from the asyncResponse. Logically this should never happen. Added to be safe.
         InstrumentationContext.get(AsyncResponse.class, Span.class).put(asyncResponse, null);
       }
-      if (asyncResponse == null || !asyncResponse.isSuspended()) {
+      if (asyncReturnValue != null) {
+        // span finished by CompletionStageFinishCallback
+        asyncReturnValue = asyncReturnValue.whenComplete(new CompletionStageFinishCallback<>(span));
```
Comment (anuraaga) [inline instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-common/src/main/java/io/opentelemetry/instrumentation/auto/jaxrs/v2_0/JaxRsAnnotationsInstrumentation.java:169]: I don't think we consistently do so in this codebase, but FWIU `handle` has much better performance than `whenComplete` (due to a JDK bug?)

https://github.com/line/armeria/issues/1426

Can we use handle here?

File: instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-testing/src/main/groovy/JaxRsHttpServerTest.groovy [hunk 2]
```diff
@@ -73,6 +104,15 @@ abstract class JaxRsHttpServerTest<S> extends HttpServerTest<S> {
     true
   }
 
+  private static boolean supportsJaxRs21() {
```
Comment (anuraaga) [inline instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-testing/src/main/groovy/JaxRsHttpServerTest.groovy:107]: If this is false normally and true when running latest deps, recommend being more explicit about it like our redisson instrumentation

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation/redisson-3.0/redisson-3.0.gradle#L21

File: instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-testing/src/main/groovy/JaxRsHttpServerTest.groovy [hunk 3]
```diff
@@ -58,6 +59,36 @@ abstract class JaxRsHttpServerTest<S> extends HttpServerTest<S> {
...
+    assumeTrue(supportsJaxRs21())
+
+    given:
+    def url = HttpUrl.get(address.resolve("/async-completion-stage")).newBuilder()
+      .addQueryParameter("action", action)
+      .build()
+    def request = request(url, "GET", null).build()
+
+    when:
+    def response = client.newCall(request).execute()
```
Comment (anuraaga) [inline instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-testing/src/main/groovy/JaxRsHttpServerTest.groovy:73]: IIUC, it's important our test makes sure a span hasn't been ended before the future is completed. How about adding some sleep in the API method, and separate out making the call, checking there are no traces (with previous code I think there would have been), than joining on the future, then checking there are traces?

**Review by anuraaga — COMMENTED** [review]
Cool, thanks

File: instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-common/src/main/java/io/opentelemetry/instrumentation/auto/jaxrs/v2_0/JaxRsAnnotationsInstrumentation.java (hunk 1 above)
Comment (mateuszrzeszutek) [reply instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-common/src/main/java/io/opentelemetry/instrumentation/auto/jaxrs/v2_0/JaxRsAnnotationsInstrumentation.java]: > Can we use handle here?

Of course -- changed. Thanks!

File: instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-testing/src/main/groovy/JaxRsHttpServerTest.groovy (hunk 2 above)
Comment (mateuszrzeszutek) [reply instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-testing/src/main/groovy/JaxRsHttpServerTest.groovy]: 👍 
That's much better, done.

File: instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-testing/src/main/groovy/JaxRsHttpServerTest.groovy (hunk 3 above)
Comment (mateuszrzeszutek) [reply instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-testing/src/main/groovy/JaxRsHttpServerTest.groovy]: Instead of adding sleep (which would make this test at least a bit flaky) I've added synchronisation using a `CyclicBarrier` -- but the principle is the same: no traces until the future is complete.

**Review by anuraaga — APPROVED** [review]
Thanks!

File: instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-jersey-2.0/jaxrs-2.0-jersey-2.0.gradle [hunk 4]
```diff
@@ -24,3 +24,7 @@ dependencies {
   testImplementation group: 'javax.xml.bind', name: 'jaxb-api', version: '2.2.3'
   testImplementation group: 'com.fasterxml.jackson.module', name: 'jackson-module-afterburner', version: '2.9.10'
 }
+
+test {
+  systemProperty 'testLatestDeps', testLatestDeps
```
Comment (iNikem) [inline instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-jersey-2.0/jaxrs-2.0-jersey-2.0.gradle:29]: Why is this needed? @anuraaga is it?

Comment (iNikem) [reply instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-jersey-2.0/jaxrs-2.0-jersey-2.0.gradle:29]: A, got it...

**Review by iNikem — APPROVED** [review]

File: instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-jersey-2.0/jaxrs-2.0-jersey-2.0.gradle (hunk 4 above)
Comment (anuraaga) [reply instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-jersey-2.0/jaxrs-2.0-jersey-2.0.gradle:29]: A, but this is a great point out. @mateuszrzeszutek if you get a chance to add a one line comment here sometime to avoid future confusion would be great!
